asdf/_display.py

```diff
@@ -82,8 +83,9 @@ def from_root_node(cls, root_identifier, root_node):
             next_nodes = []
 
             for parent, identifier, node in current_nodes:
-                if (isinstance(node, dict) or isinstance(node, list) or isinstance(node, tuple)) and id(node) in seen:
-                    info = _NodeInfo(parent, identifier, node, current_depth, recursive=True)
+                if (isinstance(node, Mapping) or isinstance(node, MutableSequence) or isinstance(node, tuple)) and id(node) in seen:
```

Comment:
I'm starting to have some reservations about crawling all subclasses of `Mapping` and `MutableSequence`.  I assume we're skipping `Sequence` here because that would match strings -- are there other types that will end up being similarly inconvenient to have split into separate nodes in the info display?

What do you think about having the user or the extension opt in each traversable class?  On an extension that might look something like:

```python
class SomeExtension:
  traversable_types = [
    "roman_datamodels.stnode.DNode",
    "roman_datamodels.stnode.LNode",
  ]
```

Reply:
Alternatively we could create a marker interface, something like:

```python
class AsdfTraversable:
  pass
```

and include that among DNode and LNode's parent classes:

```python
from asdf.extension import AsdfTraversable

class DNode(UserList, AsdfTraversable):
  # blah blah
```

I like that better from an implementation point of view, but an extension author wouldn't be able to mark classes that they don't control.

Reply:
Belatedly, I don't see that as a problem normally. It's needed here since we are building a tree of objects, and if someone else has a similar need, they can use a similar solution.

Reply:
As I try to sweep the cobwebs away on this one. Is there any reason that DNode and LNode shouldn't automatically be traversed?

As for other possibly traversable objects, we could have such objects provide a method to return whatever structure they think appropriate for their contents. Any reason not to let them control that completely? The transversing code checks for the presence of such a method, and if it exists, calls it. Should there be reference objects within that also support that interface, they can also be called, recursively. I suppose we could use this for DNode and LNodes as a uniform interface.

Reply:
I like the idea of a method, `__asdf_traverse__`?

Reply:
But how to handle classes that are implemented by a 3rd party and not necessary modifiable by asdf developers or the extension author?

Reply:
OK, I'll give it a try. That seems like a good name. I was going to go with `__asdf_the_doors_open_so_come_in_and_look_around__` but that might be a hair too long.

Reply:
I suppose I should close this and open a new PR. What should I base it off of given all the changes going on?

Reply:
As for 3rd party classes that don't implement it, tough. They don't want to be introspected, would be the simple answer. Otherwise we could implement a registry of some sort that knows about them and has code to deal with that. But not for this PR.

Reply:
I recommend branching off of master so that we can release this sooner in a 2.x version.